staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"metadata": *objectTypePtr(map[string]schema.Structural{
+							"name":         stringType,
+							"generateName": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.kind == 'Pod'",
+				"embedded.apiVersion == 'v1'",
+				"embedded.metadata.name == 'foo'",
+				"embedded.metadata.generateName == 'pickItForMe'",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata.namespace)": "undefined field 'namespace'",
+				"has(embedded.spec)":               "undefined field 'spec'",
+			},
+		},
+		{name: "embedded object with properties",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"metadata": *objectTypePtr(map[string]schema.Structural{
+							"name":         stringType,
+							"generateName": stringType,
+						}),
+						"spec": *objectTypePtr(map[string]schema.Structural{
+							"field1": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.kind == 'Pod'",
+				"embedded.apiVersion == 'v1'",
+				"embedded.metadata.name == 'foo'",
+				"embedded.metadata.generateName == 'pickItForMe'",
+				"embedded.spec.field1 == 'a'",
+			},
+		},
+		{name: "embedded object with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			valid: []string{
+				// access to metadata allowed even if not specified in schema
+				// because the accessible fields (name and generateName) are fixed
+				"embedded.kind == 'Pod'",
+				"embedded.apiVersion == 'v1'",
+				"embedded.metadata.name == 'foo'",
+				"embedded.metadata.generateName == 'pickItForMe'",
```

Comment:
you could add:
```
"embedded.metadata.namespace == 'xyz'",
```

Reply:
I'll add this to the error cases to document that it is inaccessible